File: contests/lc/3047.FindtheLargestAreaofSquareInsideTwoRectangles/solution.py
from typing import List


class Solution:
    def largestSquareArea(self, bottomLeft: List[List[int]], topRight: List[List[int]]) -> int:
        max_side = 0
        n = len(bottomLeft)
        for i in range(n):
            for j in range(i + 1, n):
                left = max(bottomLeft[i][0], bottomLeft[j][0])
                right = min(topRight[i][0], topRight[j][0])
                bottom = max(bottomLeft[i][1], bottomLeft[j][1])
                top = min(topRight[i][1], topRight[j][1])

                width = right - left
                height = top - bottom
                if width <= 0 or height <= 0:
                    continue

                side = min(width, height)
                if side > max_side:
                    max_side = side

        return max_side * max_side

File: contests/lc/3047.FindtheLargestAreaofSquareInsideTwoRectangles/solution.go
package main

func largestSquareArea(bottomLeft [][]int, topRight [][]int) int64 {
	maxSide := int64(0)
	n := len(bottomLeft)
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			left := maxInt(bottomLeft[i][0], bottomLeft[j][0])
			right := minInt(topRight[i][0], topRight[j][0])
			bottom := maxInt(bottomLeft[i][1], bottomLeft[j][1])
			top := minInt(topRight[i][1], topRight[j][1])

			width := right - left
			height := top - bottom
			if width <= 0 || height <= 0 {
				continue
			}

			side := int64(width)
			if height < width {
				side = int64(height)
			}
			if side > maxSide {
				maxSide = side
			}
		}
	}

	return maxSide * maxSide
}

func maxInt(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func minInt(a, b int) int {
	if a < b {
		return a
	}
	return b
}
